src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+   * ignore all future messages from that Subscriber, including future calls to
+   * Cancel.
+   *
+   * Publishers must support Subscribers calling Cancel() from inside
+   * Subscriber::Receive(...).
+   */
+  virtual void Cancel() = 0;
+
+  /**
+   * Requests the Publisher to send the indicated number of values to the
+   * Subscriber.
+   *
+   * Publishers must support Subscribers calling Request(Demand) from inside
+   * Subscriber::Receive(Subscription), but Subscribers should avoid calling
+   * Request(Demand) inside Subscriber::Receive(Input). Instead, they should
+   * send additional Demand via the return value of Subscriber::Receive(Input)
+   * (to help prevent infinite recursion).
+   */
+  virtual void Request(Demand demand) = 0;
+};
+
+/**
+ * Type representing a message from a Publisher to a Subscriber indicating that
+ * the Subscriber will no longer receive any further messages.
+ */
+class Completion {
+ public:
+  /** Returns an instance that signals successful completion. */
+  static Completion Finished() { return Completion(); }
+
+  /**
+   * Returns an instance that signals failure, described by the given
+   * exception.
+   */
+  static Completion Failure(std::exception_ptr e) { return Completion(e); }
+
+  bool IsFinished() const { return failure_ == nullptr; }
+
+  /** Returns the exception if a failure and a null pointer otherwise. */
+  std::exception_ptr failure() const { return failure_; }
+
+ private:
+  explicit Completion(std::exception_ptr e = nullptr) : failure_(e) {}
+
+  std::exception_ptr failure_;
+};
+
+/**
+ * Interface for objects that consume values from a Publisher.
+ *
+ * Unless otherwise specified by the concrete implementation, external
+ * synchronization must be used to avoid concurrent calls the Subscriber
+ * interface from different threads.
+ */
+template <typename T>
+class Subscriber {
+ public:
+  /** The type of the values that this Subscriber consumes. */
+  using Input = T;
+
+  virtual ~Subscriber() = default;
+
+  /**
+   * The first signal that a Subscriber receives from a Publisher, passing the
+   * Subscription that the Subscriber can use to control the flow of values.
+   *
+   * A Subscriber may only have one Publisher. If it somehow receives more than
+   * one Subscription, it should call Subscription::Cancel() on any instances
+   * received after the first.
+   *
+   * A Subscriber is explictly allowed to demand values synchronously from
+   * within its implementation of this method.
+   */
+  virtual void Receive(std::shared_ptr<Subscription> subscription) = 0;
+
+  /**
+   * Transmits a value from the Publisher to this Subscriber.
+   *
+   * A Subcriber should never receive more calls to this method than the total
+   * Demand it has requested from its publisher. Subscribers should only demand
+   * more elements from within this method via its return value.
+   */
+  virtual Demand Receive(Input element) = 0;
```

Comment:
this may cause an infinite subscription. If need more data, it's better to have a new subscription. In this way, the subscriber is not responsible for accounting but only for receiving published values.

Reply:
We’re hewing to the design of the Combine API published in Swift.

This will help with familiarity for contributors more familiar with Swift, and in case we migrate this codebase to Swift in future

Reply:
Thanks for letting me know what's the motivation behind.